lib/task/migrate/migrations/arMigration0184.class.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/*
+ * This file is part of the Access to Memory (AtoM) software.
+ *
+ * Access to Memory (AtoM) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Access to Memory (AtoM) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Add tables, taxonomy, and terms that relate to accession events
+ *
+ * @package    AccesstoMemory
+ * @subpackage migration
+ */
+class arMigration0184
+{
+  const
+    VERSION = 184, // The new database version
+    MIN_MILESTONE = 2; // The minimum milestone required
+
+  public function up($configuration)
+  {
+    // Add accession_event table
+    $sql = <<<sql
+
+CREATE TABLE IF NOT EXISTS `accession_event`
+(
+        `id` INTEGER  NOT NULL,
+        `type_id` INTEGER,
+        `accession_id` INTEGER,
+        `date` DATE,
+        `source_culture` VARCHAR(16)  NOT NULL,
+        PRIMARY KEY (`id`),
+        CONSTRAINT `accession_event_FK_1`
+                FOREIGN KEY (`id`)
+                REFERENCES `object` (`id`)
+                ON DELETE CASCADE,
+        INDEX `accession_event_FI_2` (`type_id`),
+        CONSTRAINT `accession_event_FK_2`
+                FOREIGN KEY (`type_id`)
+                REFERENCES `term` (`id`)
+                ON DELETE SET NULL,
+        INDEX `accession_event_FI_3` (`accession_id`),
+        CONSTRAINT `accession_event_FK_3`
+                FOREIGN KEY (`accession_id`)
+                REFERENCES `accession` (`id`)
+                ON DELETE CASCADE
+)Engine=InnoDB;
+
+CREATE TABLE `accession_event_i18n`
+(
+        `agent` VARCHAR(255),
+        `id` INTEGER  NOT NULL,
+        `culture` VARCHAR(16)  NOT NULL,
+        PRIMARY KEY (`id`,`culture`),
+        CONSTRAINT `accession_event_i18n_FK_1`
+                FOREIGN KEY (`id`)
+                REFERENCES `accession_event` (`id`)
+                ON DELETE CASCADE
+)Engine=InnoDB;
+
+sql;
+
+    // Add accession event type taxonomy if it doesn't exist
+    $sql = "SELECT * FROM ".QubitTaxonomyI18n::TABLE_NAME."
+              WHERE name='Accession event type'
+              AND culture='en'
+              AND id=?";
+
+    if (null == QubitPdo::fetchOne($sql, array(QubitTaxonomy::ACCESSION_EVENT_TYPE_ID)))
```

Comment:
Optimizations getting into your head, I know the feeling lately, hehehe. I think it would have been okay with `QubitTaxonomy::getById()`. Also, really unlikely but the name may have changed, I'd use only the id to query over the taxonomy table.

Reply:
And similar for the terms below ;)

Reply:
Yeah, you're right... `getById` makes sense for QubitTaxonomy given users can add to it. For the terms, however, wouldn't there be, without a name check, the risk that a user had added a custom term using the ID?

Reply:
Oh, I didn't think about that, great point @mcantelon. Moreover, without being protected terms, the name could have changed. Should they be protected? We're getting quite edgy in here knowing that this will be partially cherry picked to a 2.5.4 version, so it's probably okay as it was.

@sbreker, if we don't protect the terms in the end, we should at least notify that the term names shouldn't be changed for the 2.6 upgrade to work properly.

Reply:
@mcantelon, you could add an extra condition over the `taxonomy_id` to those term queries to reduce a bit the risk.

Also, do you know if we need the same checks on arMigration0179? (@sbreker and @djjuhasz too)

https://github.com/artefactual/atom/blob/qa/2.6.x/lib/task/migrate/migrations/arMigration0179.class.php

Reply:
Yeah, might be good to add those checks to arMigration0179. I'll make a separate PR for that.

I've made the two terms added by migration 184 protected and added taxonomy_id checks for both.